src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs

```diff
@@ -22,7 +29,54 @@ public static void Initialize()
             ImageAnalytics.Initialize();
         }
 
+        private static unsafe ISchema GetModelSchema(IExceptionContext ectx, TFGraph graph)
+        {
+            long pos = 0;
+            TF_Operation oper;
+            var res = new List<KeyValuePair<string, ColumnType>>();
+            while ((oper = TFGraph.TF_GraphNextOperation(graph.handle, &pos)) != IntPtr.Zero)
+            {
+                var name = TFGraph.TF_OperationName(oper);
+                var type = TFGraph.TF_OperationOpType(oper);
+                var numOutputs = TFGraph.TF_OperationNumOutputs(oper);
+                if (numOutputs != 1)
+                    continue;
+
+                var numInputs = TFGraph.TF_OperationNumInputs(oper);
+                if (numInputs == 0)
+                    continue;
+
+                var tfType = TFGraph.TF_OperationOutputType(new TFOutput(graph[name]));
+                var mlType = Tf2MlNetTypeOrNull(tfType);
+                if (mlType == null)
+                    continue;
+
+                var shape = graph.GetTensorShape(new TFOutput(graph[name]));
+                var shapeArray = shape.ToIntArray();
+                var columnType = Utils.Size(shapeArray) > 0 && shapeArray.Skip(1).All(x => x > 0) ?
+                    new VectorType(mlType, shapeArray[0] > 0 ? shapeArray : shapeArray.Skip(1).ToArray())
+                    : new VectorType(mlType);
+                res.Add(new KeyValuePair<string, ColumnType>(name, columnType));
+            }
+            return new SimpleSchema(ectx, res.ToArray());
+        }
+
+        public static ISchema GetModelSchema(IExceptionContext ectx, string modelFile)
+        {
+            var bytes = File.ReadAllBytes(modelFile);
+            var session = LoadTFSession(ectx, bytes, modelFile);
```

Comment:
>LoadTFSession [](start = 26, length = 13)

What about models that are not frozen? I assume it will be there once @Abhishek changes are there, right?

Reply:
Yes. I assume that LoadTFSession will take care of the logic to decide which kind of model to load.

---
In reply to: [216509938](https://github.com/dotnet/machinelearning/pull/862#discussion_r216509938) [](ancestors = 216509938)